### ISSUE TITLE
Feature Request: keepdims option for tf.boolean_mask()

### ISSUE THREAD

**Issue by datwelk** [tensorflow/tensorflow#18238]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: N/A
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/A
- **TensorFlow installed from (source or binary)**: N/A
- **TensorFlow version (use command below)**: ('v1.6.0-0-gd2e24b6039', '1.6.0')
- **Python version**: Python 2.7.14
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

### Describe the problem
`tf.boolean_mask()` currently grabs individual elements from the input tensor according to the shape and content of the boolean mask argument (see https://www.tensorflow.org/api_docs/python/tf/boolean_mask). The returning tensor's dimensionality is always lower than or equal to the input tensor. It would be very useful if `tf.boolean_mask()` included an additional, optional boolean argument `keepdims`. If `keepdims=False`, the behavior of the method is unchanged. If `keepdims=True`, the output tensor would have the same dimensionality as the input tensor. This would allow 'filtering out' of certain values from the input, while keeping its overall shape as much as possible (see example below).

### Source code / logs
```
tensor = [[1, 2], [3, 4], [5, 6]]
mask = np.array([[True, False],[False, True],[False, True]])
boolean_mask(tensor, mask, keepdims=False)  # [1, 4, 6]
boolean_mask(tensor, mask, keepdims=True) # [[1],[4],[6]]

mask = np.array([[True, False],[False, True],[False, False]])
boolean_mask(tensor, mask, keepdims=False)  # [1, 4]
boolean_mask(tensor, mask, keepdims=True) # [[1],[4]]
```

**Comment by burglarhobbit** [comment]
There's a little flaw in your logic, if the tensor and mask is like this: 
```
tensor = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
mask = np.array([[True, False, True], [False, False, True], [True, True, True]])
boolean_mask(tensor, mask, keepdims=False) # [1, 3, 6, 7, 8, 9]
boolean_mask(tensor, mask, keepdims=True, pad_val=0) # [[1, 3, 0], [6, 0, 0], [7, 8, 9]] 
``` 
then expected output according to you should be `[[1, 3], [6], [7, 8, 9]]` but this will give ShapeError since TensorFlow doesn't allow flexible shapes like Python lists do. So what I suggest is along with `keepdims` argument, there should be a 'pad_val` argument where you mention the value of the padding that is going to happen at the end.

**Comment by bignamehyp** [comment]
@langmore can you please take a look?

I think boolean_mask(tensor, mask) is designed to implement the numpy tensor[mask] function. And you can work around this by using cond_vars.

**Comment by lc82111** [comment]
I run into the same problem, and need keepdims feature for tf.boolean_mask().

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by panfengli** [comment]
I am wondering after `tf.boolean_mask`, how to recover the predicted results as a 2D tensor from pred (1D tensor) and mask (2D array with elements of 'True' or 'False'), where False corresponds to 0.

```
mask = labels > 0
labels = tf.boolean_mask(labels, mask)
heatmaps = tf.boolean_mask(heatmaps, mask)

scores = tf.nn.softmax(logits=heatmaps, labels=labels)
pred = tf.argmax(scores)
```

@langmore 

Thanks!

**Comment by ppwwyyxx** [comment]
I think this is still not a well-defined feature request. There aren't even some examples of (input, output) pairs at all. The input, output pairs given by in the original issue are already proven unsound by @burglarhobbit.

If one just wants to replace False entries by another element (which seems to be what @panfengli needs) then you should use `tf.where`.

**Comment by panfengli** [comment]
For example,

```
img = tf.constant(([1,2,3], [4,5,6], [7,8,9]))
mask = tf.where(img>5)
new_img = tf.zeros_like(img)
```
I would like new_img to be 
```
[[0, 0, 0],
[0, 5, 6],
[7, 8, 9]]
```

**Comment by ppwwyyxx** [comment]
@panfengli  `tf.where(img>5, img, tf.zeros_like(img))` does what you need.

**Comment by panfengli** [comment]
Okay, I got it. Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Santosh-Gupta** [comment]
> then expected output according to you should be `[[1, 3], [6], [7, 8, 9]]` but this will give ShapeError since TensorFlow doesn't allow flexible shapes like Python lists do. So what I suggest is along with `keepdims` argument, there should be a `pad_val` argument where you mention the value of the padding that is going to happen at the end.

Is this still the case now that Ragged tensors have been released?

**Comment by edloper** [comment]
In PR 1366762 we added a new "tf.ragged.boolean_mask" op which preserves mask dimensions.

**Comment by jamie0725** [comment]
Hi @edloper, what if I want to keepdim but also need a valid tensor by adding padding, like what @burglarhobbit said:
```
tensor = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
mask = np.array([[True, False, True], [False, False, True], [True, True, True]])
boolean_mask(tensor, mask, keepdims=False) # [1, 3, 6, 7, 8, 9]
boolean_mask(tensor, mask, keepdims=True, pad_val=0) # [[1, 3, 0], [6, 0, 0], [7, 8, 9]] 
```
